include/xsparse/levels/range.hpp

```diff
@@ -0,0 +1,79 @@
+#ifndef XSPARSE_LEVELS_RANGE_HPP
+#define XSPARSE_LEVELS_RANGE_HPP
+
+#include <utility>
+#include <xsparse/util/base_traits.hpp>
+#include <xsparse/level_capabilities/coordinate_iterate.hpp>
+#include <tuple>
+
+namespace xsparse
+{
+    namespace levels
+    {
+        template <class LowerLevels, class IK, class PK, class OffsetContainer>
+        class range;
+
+        template <class... LowerLevels, class IK, class PK, class OffsetContainer>
+        class range<std::tuple<LowerLevels...>, IK, PK, OffsetContainer>
+            : public level_capabilities::coordinate_value_iterate<range,
+                                                                  std::tuple<LowerLevels...>,
+                                                                  IK,
+                                                                  PK,
+                                                                  OffsetContainer>
+        {
+            using BaseTraits
+                = util::base_traits<range, std::tuple<LowerLevels...>, IK, PK, OffsetContainer>;
+
+        public:
+            range(IK size_N, IK size_M)
+                : m_size_N(std::move(size_N))
+                , m_size_M(std::move(size_M))
```

Comment:
```suggestion
                : m_size_M(std::move(size_M))
                , m_size_N(std::move(size_N))
```

For optimization reasons, all fields must be initialized in the same order in which they're declared, and declared in the same order in which they're used. Smaller fields should go on top and larger ones on the bottom.

Reply:
This applies to all constructors and classes/structs.

Reply:
The paper uses NxM (rows x columns) convention while referring to the dimensions of the matrix, would it be best to change the declaration and initializations to follow this convention?

Reply:
That works, yes.